rpcs3/rpcs3qt/skylander_dialog.cpp

```diff
@@ -642,273 +531,313 @@ u16 skylander_dialog::skylander_crc16(u16 init_value, const u8* buffer, u32 size
+		sky_file.close();
+
+		last_skylander_path = QFileInfo(file_path).absolutePath() + "/";
+		accept();
+	});
 
-		// Type 1
-		reinterpret_cast<le_t<u16>&>(decrypted_header[0xE]) = 5;
-		reinterpret_cast<le_t<u16>&>(decrypted_header[0xE]) = skylander_crc16(0xFFFF, decrypted_header.data(), 16);
+	connect(btn_cancel, &QAbstractButton::clicked, this, [&]() { reject(); });
 
-		set_block(active, decrypted_header);
-		set_block(active + 9, sub_header);
-	}
+	connect(co_compl, QOverload<const QString&>::of(&QCompleter::activated),[=](const QString& text)
+	{
+		combo_skylist->setCurrentText(text);
+		combo_skylist->setCurrentIndex(combo_skylist->findText(text));
+	});	
 }
 
-bool skylander_dialog::is_initialized()
+QString skylander_creator_dialog::get_file_path() const
 {
-	std::lock_guard lock(g_skylander.sky_mutex);
-	for (u32 index = 1; index < 0x10; index++)
-	{
-		for (u32 subdex = 0; subdex < (0x30 / sizeof(u64)); subdex++)
-		{
-			if (reinterpret_cast<const u64&>(g_skylander.sky_dump[(index * 0x40) + (subdex * sizeof(u64))]) != 0)
-			{
-				return true;
-			}
-		}
-	}
-	return false;
+	return file_path;
 }
 
-void skylander_dialog::new_skylander()
+skylander_dialog::skylander_dialog(QWidget* parent)
+    : QDialog(parent)
 {
-	const QString file_path = QFileDialog::getSaveFileName(this, tr("Create Skylander File"), cur_sky_file_path, tr("Skylander Object (*.sky);;"));
-	if (file_path.isEmpty())
-		return;
+	setWindowTitle(tr("Skylanders Manager"));
+	setObjectName("skylanders_manager");
+	setAttribute(Qt::WA_DeleteOnClose);
+	setMinimumSize(QSize(700, 200));
+
+	QVBoxLayout* vbox_panel = new QVBoxLayout();
 
-	if (g_skylander.sky_file)
-		g_skylander.sky_file.close();
+	auto add_line = [&](QVBoxLayout* vbox) {
+		QFrame* line = new QFrame();
+		line->setFrameShape(QFrame::HLine);
+		line->setFrameShadow(QFrame::Sunken);
+		vbox->addWidget(line);
+	};
 
-	g_skylander.sky_file.open(file_path.toStdString(), fs::read + fs::write + fs::create);
-	if (!g_skylander.sky_file)
-		return;
+	QGroupBox* group_skylanders = new QGroupBox(tr("Active Portal Skylanders:"));
+	QVBoxLayout* vbox_group     = new QVBoxLayout();
 
-	cur_sky_file_path = file_path;
+	QPushButton* clear_btns[UI_SKY_NUM]  = {};
+	QPushButton* create_btns[UI_SKY_NUM] = {};
+	QPushButton* load_btns[UI_SKY_NUM]   = {};
 
+	for (auto i = 0; i < UI_SKY_NUM; i++)
 	{
-		std::lock_guard lock(g_skylander.sky_mutex);
-		memset(g_skylander.sky_dump, 0, 0x40 * 0x10);
+		QHBoxLayout* hbox_skylander = new QHBoxLayout();
+		QLabel* label_skyname       = new QLabel(QString(tr("Skylander %1")).arg(i + 1));
+		edit_skylanders[i]          = new QLineEdit();
+		edit_skylanders[i]->setEnabled(false);
 
-		// Set the block permissions
-		reinterpret_cast<le_t<u32>&>(g_skylander.sky_dump[0x36]) = 0x690F0F0F;
-		for (u32 index = 1; index < 0x10; index++)
-		{
-			reinterpret_cast<le_t<u32>&>(g_skylander.sky_dump[(index * 0x40) + 0x36]) = 0x69080F7F;
-		}
+		clear_btns[i]  = new QPushButton(tr("Clear"));
```

Comment:
I don't see a reason for these arrays.
Just use a singular QPushButton and move the connect from the loop below to this loop
```suggestion
		QPushButton* button  = new QPushButton(tr("Clear"));
```

Reply:
It was more a stylistic choice I guess, to have the connects after the window is designed but changed.